source/_posts/2017-11-02-kafka-streams-dsl-vs-processor-api.markdown

```diff
@@ -0,0 +1,542 @@
+ClientKey("bob"), EvPv("ev0", "show header", "pv0", "/")
+ClientKey("bob"), EvPv("ev1", "show ads", "pv0", "/")
+ClientKey("bob"), EvPv("ev2", "show recommendation", "pv0", "/")
+ClientKey("bob"), EvPv("ev3", "click recommendation", "pv0", "/")
+
+// Events from offer page, somehow incomplete due to streaming semantics limitations
+ClientKey("bob"), EvPv("ev0", "show header", None, None)
+ClientKey("bob"), EvPv("ev1", "show ads", "pv1", "/offer?id=1234")
+ClientKey("bob"), EvPv("ev2", "add to cart", None, None)
+```
+
+## Kafka Stream DSL
+
+Now we are ready to implement above use case with recommended Kafka Streams DSL. 
+The code could be optimized but I would like present canonical way of using DSL
+without thinking about DSL internals.
+All examples are implemented using the latest Kafka Streams 1.0.0 version.
+
+Create two input streams for page views and events 
+connected to "clickstream.events" and "clickstream.page_views" Kafka topics.
+
+``` scala
+val builder = new StreamsBuilder()
+
+val evStream = builder.stream[ClientKey, Ev]("clickstream.events")
+val pvStream = builder.stream[ClientKey, Pv]("clickstream.page_views")
+```
+
+Repartition topics by client and page view identifiers `PvKey` 
+as a prerequisite to join events with page view. 
+
+``` scala
+case class PvKey(clientId: ClientId, pvId: PvId)
+
+val evToPvKeyMapper: KeyValueMapper[ClientKey, Ev, PvKey] =
+  (clientKey, ev) => PvKey(clientKey.clientId, ev.pvId)
+
+val evByPvKeyStream = evStream.selectKey(evToPvKeyMapper)
+
+val pvToPvKeyMapper: KeyValueMapper[ClientKey, Pv, PvKey] =
+  (clientKey, pv) => PvKey(clientKey.clientId, pv.pvId)
+
+val pvByPvKeyStream = pvStream.selectKey(pvToPvKeyMapper)
+```
+
+Join event with page view streams by selected `PvKey`, 
+left join is used because we are interested also in events without matched page view.
+The join window duration is set to 10 minutes.
+It means, that Kafka Streams will look for messages in "this" and "other" side of the join 
+10 minutes in the past and 10 minutes in the future (using event time, not wall-clock time).
+Because we are not interested in late events out of defined window, 
+the retention is defined 2 times longer than window.
+If you are interested why 1 millis needs to be added to the retention, 
+please ask Kafka Streams architects not me ;)
+
+``` scala
+val evPvJoiner: ValueJoiner[Ev, Pv, EvPv] = { (ev, pv) =>
+  if (pv == null) {
+    EvPv(ev.evId, ev.value, None, None)
+  } else {
+    EvPv(ev.evId, ev.value, Some(pv.pvId), Some(pv.value))
+  }
+}
+
+val joinWindowDuration = 10 minutes
+
+val joinRetention = joinWindowDuration.toMillis * 2 + 1
+val joinWindow = JoinWindows.of(joinWindowDuration.toMillis).until(joinRetention)
+
+val evPvStream = evByPvKeyStream.leftJoin(pvByPvKeyStream, evPvJoiner, joinWindow)
+```
+
+Now it's time to fight with duplicated enriched events, 
+joined stream needs to be repartitioned again by compound key of client, page view and event identifiers `EvPvKey`.
```

Comment:
może: compound key `EvPvKey` composed of client, page view and event identifiers.